sem/tree: mark CastExpr as operatorExpr

yuzefovich opened the pull request:
CastExpr now implements operatorExpr interface which is a marker
interface to know whether we should be surrounding the expression with
parenthesis when formatting it.

Fixes: #44137.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44261)
<!-- Reviewable:end -->

yuzefovich commented:
The comment in the code suggested that the type checking must have succeeded by the point where an assertion was placed, but it seems to me that there cases when it was not correct. Namely, if use the logic test with `local` config, the error doesn't occur at all, so I switched to `fakedist` config to repro.

I might be totally off here.

knz reviewed (changes requested):
> it seems to me that in distributed setting such errors are expected to occur

wait what? Why? All type error must be caught during planning. If they are not, there is an architectural mistake that needs to be fixed. 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44261#-:-LzHadi8-4vQhrOc20Kj:b-zc977k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/typing, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44261#-LzHauZY9lguERCbSZNM:-LzHauZY9lguERCbSZNN:b-7eqgwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/584ef227715a79b3a6ac6fb1f7ee1f003aa57c65/pkg/sql/logictest/testdata/logic_test/typing#L206)):*
> ```
> SELECT * FROM t44137 WHERE 1 > -(('1a' COLLATE en)::INT)
> ```

The output of EXPLAIN reveals what the problem is:

```
root@127.0.0.1:25137/movr> explain(verbose) SELECT * FROM t44137 WHERE 1 > -(('1a' COLLATE en)::INT);
  tree |    field    |         description          | columns | ordering
-------+-------------+------------------------------+---------+-----------
       | distributed | true                         |         |
       | vectorized  | false                        |         |
  scan |             |                              | (c0)    |
       | table       | t44137@primary               |         |
       | spans       | ALL                          |         |
       | filter      | (-'1a' COLLATE en::INT8) < 1 |         |
```

The problem is that the pretty-printing has removed the necessary parentheses. The resulting expression is different. This is a distsql serialization bug.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sem/tree: mark CastExpr as operatorExpr

CastExpr now implements operatorExpr interface which is a marker
interface to know whether we should be surrounding the expression with
parenthesis when formatting it.

Release note: None

yuzefovich reviewed (commented):
I agree, I was off in the first attempt.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44261#-:-LzIZDGV4gmwGv1vNqJo:bk0xiew)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/logictest/testdata/logic_test/typing, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44261#-LzHauZY9lguERCbSZNM:-LzIZJGm8GhcswUUlKZH:b-3ueafq) ([raw file](https://github.com/cockroachdb/cockroach/blob/584ef227715a79b3a6ac6fb1f7ee1f003aa57c65/pkg/sql/logictest/testdata/logic_test/typing#L206)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

> ```
> SELECT * FROM t44137 WHERE 1 > -(('1a' COLLATE en)::INT)
> ```

The output of EXPLAIN reveals what the problem is:

```
root@127.0.0.1:25137/movr> explain(verbose) SELECT * FROM t44137 WHERE 1 > -(('1a' COLLATE en)::INT);
  tree |    field    |         description          | columns | ordering
-------+-------------+------------------------------+---------+-----------
       | distributed | true                         |         |
       | vectorized  | false                        |         |
  scan |             |                              | (c0)    |
       | table       | t44137@primary               |         |
       | spans       | ALL                          |         |
       | filter      | (-'1a' COLLATE en::INT8) < 1 |         |
```

The problem is that the pretty-printing has removed the necessary parentheses. The resulting expression is different. This is a distsql serialization bug.
</blockquote></details>

Indeed, thanks for the pointer. Distsql reuses code in `sem/tree` to format the expressions, and step-by-step debugging revealed that we didn't put `CastExpr` in parenthesis. I'm not sure if that is desired, but it solves the internal error. I don't know the semantics here at all though, so again, maybe the problem is somewhere else.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I think this is good now. However it is a user-visible change (bug fix) and thus needs a release note. It will also deserve a backport.

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44261#-:-LzIaaqI7u1Hdie0rYdN:bicn4mk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Ouch, this change broke a lot of things :/

Do you think that is expected? Maybe we should introduce something like "parenthesized cast expr" (the cast expression that was in parenthesis in the original query) and only those would be printed in parethesis?

knz commented:
ouch indeed.
I'm sorry this is off hours for me and I'm not avaiable.
Can you grab someone closer eg @otan or even @RaduBerinde or  @nvanbenschoten  and seek their advice instead.

yuzefovich commented:
The issue has been somehow resolved, so closing this PR.
